core/src/desul/atomics/Compare_Exchange_ScopeCaller.hpp

```diff
@@ -0,0 +1,43 @@
+/*
+Copyright (c) 2019, Lawrence Livermore National Security, LLC
+and DESUL project contributors. See the COPYRIGHT file for details.
+Source: https://github.com/desul/desul
+
+SPDX-License-Identifier: (BSD-3-Clause)
+*/
+
+#ifndef DESUL_ATOMICS_COMPARE_EXCHANGE_SCOPECALLER_HPP_
+#define DESUL_ATOMICS_COMPARE_EXCHANGE_SCOPECALLER_HPP_
+#include "desul/atomics/Common.hpp"
+
+namespace desul {
+
+template <class MemoryOrder>
+DESUL_INLINE_FUNCTION void atomic_thread_fence(MemoryOrder, MemoryScopeCaller) {}
+
+#define DESUL_ATOMIC_EXCHANGE_SCOPECALLER(MEMORY_ORDER)               \
+  template <typename T>                                               \
+  DESUL_INLINE_FUNCTION T atomic_exchange(                            \
+      T* dest, T value, MEMORY_ORDER, MemoryScopeCaller) {            \
+    T return_val = *dest;                                             \
+    *dest = value;                                                    \
+    return return_val;                                                \
+  }                                                                   \
+                                                                      \
+  template <typename T>                                               \
+  DESUL_INLINE_FUNCTION T atomic_compare_exchange(                    \
+      T* dest, T compare, T value, MEMORY_ORDER, MemoryScopeCaller) { \
+    T current_val = *dest;                                            \
+    if (current_val == compare) *dest = value;                        \
+    return current_val;                                               \
+  }
+
+DESUL_ATOMIC_EXCHANGE_SCOPECALLER(MemoryOrderSeqCst)
+DESUL_ATOMIC_EXCHANGE_SCOPECALLER(MemoryOrderAcqRel)
+DESUL_ATOMIC_EXCHANGE_SCOPECALLER(MemoryOrderRelease)
+DESUL_ATOMIC_EXCHANGE_SCOPECALLER(MemoryOrderAcquire)
+DESUL_ATOMIC_EXCHANGE_SCOPECALLER(MemoryOrderRelaxed)
```

Comment:
Why can't this be templated on the memory order?

Reply:
```
C:\projects\source\core\src\Kokkos_Atomics_Desul_Wrapper.hpp(226,17): error C2668: 'desul::atomic_compare_exchange': ambiguous call to overloaded function [C:\projects\source\build\core\src\kokkoscore.vcxproj]
C:\projects\source\core\src\desul/atomics/Compare_Exchange_MSVC.hpp(144,50): message : could be 'int desul::atomic_compare_exchange<T,desul::MemoryScopeCaller>(T *const ,T,T,desul::MemoryOrderRelaxed,MemoryScope)' [C:\projects\source\build\core\src\kokkoscore.vcxproj]
          with
          [
              T=int,
              MemoryScope=desul::MemoryScopeCaller
          ]
C:\projects\source\core\src\desul/atomics/Compare_Exchange_ScopeCaller.hpp(28,25): message : or       'T desul::atomic_compare_exchange<T,desul::MemoryOrderRelaxed>(T *,T,T,MemoryOrder,desul::MemoryScopeCaller)' [C:\projects\source\build\core\src\kokkoscore.vcxproj]
```

Reply:
The above error is for the Serial only build with MSVC and comes from the fact that the MSVC compare exchange is templated on Scope, but explicitly instantiated for various memory orders and thus both variants have one template parameter.